src/hypothesis/internal/validation.py

```diff
@@ -64,6 +65,8 @@ def check_valid_bound(value, name):
     """
     if value is None or isinstance(value, integer_types + (Rational,)):
         return
+    if not isinstance(value, (Real, decimal.Decimal)):
```

Comment:
Note: Given that I don't know how `math.isnan` works and the propensity for duck typing in Python, I'm not completely convinced that this doesn't break some things that worked before. (Non-Real/Decimal types for which `isnan` works.)

I haven't thought about this hard and this seems pretty nitpicky, but I wanted to leave a comment here on it anyway.

Reply:
@Zac-HD oh, I think I understand this better now and can give you concrete examples of stuff that worked before but doesn't work with your code:

```
In [12]: class MyStupidNumber:
    ...:     def __init__(self, x):
    ...:         self.x = x
    ...:     def __float__(self):
    ...:         return self.x
    ...:
    ...:

In [13]: stupid_number = MyStupidNumber(1.5)

In [14]: float(stupid_number)
Out[14]: 1.5

In [15]: math.isnan(stupid_number)
Out[15]: False

In [20]: isinstance(stupid_number, Real)
Out[20]: False
```

(`math.isnan` works iff it can cast to float, I think.)

I don't know how important it is (probably not a lot of people out in the wild are using weird numbers that aren't Reals or Decimals but can still cast to float).

Maybe try `math.isnan` and if you catch a `TypeError`, *then* give the `InvalidArgument`?